### ISSUE TITLE
REST error reporting following exceptions

### ISSUE THREAD

**Issue by cdjackson** [eclipse/smarthome#1088]
With the latest OH2 (from this afternoon, so I think it probably includes the latest ESH changes) I get exceptions when requesting GET rest/things -:

```json
{
  "error": {
    "http-code": 500,
    "exception": {
      "class": "java.lang.NullPointerException"
    }
  }
}
```

While this reporting might be great for a UI in that we can report something more useful than just error code 500 (although in reality, not much more useful) there's no way that I've found to work out the root of the problem?  Are these errors reported somewhere or are they swallowed up and all we get is this error on the REST interface - it's just not very useful for debugging.

I have a feeling there's something wrong with thing labels since I'm trying to implement the new label support in HABmin and I may have saved a null label or something.

**Comment by cdjackson** [comment]
Ok, the issue still stands, as I think better reporting is a good idea, but I found that in this case if I copied the mapdb database from my production system to eclipse I could debug this...

There's a bug with the handling of the thing labels - if there's no label, and no linked item, then this is the result.

```java
this.label = thingDTO.label != null ? thingDTO.label : item.label;
```

I'll look to produce a fix...

**Comment by kaikreuzer** [comment]
Fixed through https://github.com/eclipse/smarthome/pull/1089

**Comment by cdjackson** [comment]
Hi Kai,
This issue wasn't directly related to the #1089 - my point really is that debugging errors that occur in the REST interface is nigh on impossible as nothing seems to be logged anywhere. All we get is the single line saying "NullPointerException" which is returned to the REST interface.

Can the exceptions also be logged somewhere?  It seems they are just swallowed?

Chris

**Comment by cdjackson** [comment]
Apologies for the confusion :)

**Comment by kaikreuzer** [comment]
Apologies for not reading it carefully enough :-)
I think in general it should be the REST resource implementation, which should make sure that exceptions are handled (and logged) appropriately.
The [RuleResource](https://github.com/eclipse/smarthome/blob/master/bundles/automation/org.eclipse.smarthome.automation.rest/src/main/java/org/eclipse/smarthome/automation/rest/internal/RuleResource.java#L109) for example does this.
We could also move the logging to the `JSONResponse.createErrorResponse` method, but we still would have to make sure that this is used in case of exceptions.

**Comment by cdjackson** [comment]
The exception gets swallowed by AbstractJavaResourceMethodDispatcher. Short of changing every resource to log the exception, or implementing an abstract resource that all others, I'm not sure how this would be fixed...

Adding code to every resource is not nice - there's dozens of them already.

**Comment by kaikreuzer** [comment]
> Adding code to every resource is not nice - there's dozens of them already.

Well, if all of them have bugs, all of them need to be fixed!
And throwing exceptions up the HTTP stack instead of returning a proper (error) response, is imho a bug.

**Comment by cdjackson** [comment]
This can be closed with #1112
